### ISSUE TITLE
vmware-iso `output_directory` is ignored

### ISSUE THREAD

**Issue by axeljov** [hashicorp/packer#6941]
- Packer version: 1.3.2
- Host platform: ESXi 6.0

When using the vmware-iso builder with the remote type set to `esx5`, Packer seems to be ignoring the `output_directory` variable (https://www.packer.io/docs/builders/vmware-iso.html#output_directory) and just builds the VM on the root of the datastore. After doing a bit of testing with different packer versions I noticed this issue started happening with packer 1.2.5+, it does not happen on any previous versions. The build is successful, but it is always built on the root of the datastore.

**Comment by ajovedel** [comment]
@SwampDragons I found the code snippet that is causing this issue and managed to fix it by commenting it out, but I am not sure if there is other functionality that might be breaking by removing this. (For me it works perfect).

The `output_directory` variable is defined as `the path to the directory where the resulting virtual machine will be created.` but the following code snippet makes it so the VM will always be created on the root of the datastore if it is a remote build (e.g. esx5). https://github.com/hashicorp/packer/blob/master/builder/vmware/iso/builder.go#L53

The `ouput_directory` variable is then only used for exporting purposes as shown in line 177, but that is not what is described in the `ouput_directory` documentation. Please let me know if removing lines 52-54 makes sense and I will make a pull request.

**Comment by SwampDragons** [comment]
Thanks. I'll take a look today.

**Comment by SwampDragons** [comment]
This is kind of a duplicate of 4051; we definitely need to figure out how to set this. It may involve creating a new "remote_output_directory" option.

**Comment by azr** [comment]
The change on this code snippet is a 2yo PR; https://github.com/hashicorp/packer/pull/4538/files

Before the PR it was :
https://github.com/hashicorp/packer/blob/0e80d98805c7756059d0714f9f623f50f56d2992/builder/vmware/iso/builder.go#L259-L261

To quote Matt:
> It looks like previously omitting a format will implicitly skip export.

I think a valid fix could be to check if `VMName` is not set before defaulting it; but I might be wrong.
I'll do a quick PR anyway.

**Comment by SwampDragons** [comment]
The real problem here is that we have the remote and local builders in the same file, and it causes all kinds of wacky problems with how we're setting things.  I thing issues like this would be less of a problem if we separated the local and remote vmware builders.

**Comment by arizvisa** [comment]
We're going to have a crazy time separating these hacks and options once a vsphere builder gets developed/merged. Just doing a line count of tests against "esx5" or "remote_driver" shows a lot of logic specific to the driver backend that's being used.

**Comment by arizvisa** [comment]
Do y'all think it's worth it to just split up this logic across the "builder" boundary into something like a vsphere-iso/vsphere-vmx (or vmware-esx5-iso/vmware-esx5-vmx) builder? Or just split it across a file boundary? So in that way, the logic of the local drivers and remote drivers are completely separate which also reduces the complexity of the configuration options (where some options are esx-only, and some are vmware-only). I totally don't mind doing this if that's the case since it's literally a copy-paste job, but...

I've been on a vsphere kick tonight looking through all issues that are related to the esx5 driver and the vsphere post-processor, and there's quite a number of them (mostly name, networking, etc). Seeing SwampDragons mention the separation of local and remote builders for VMware just made me think about the work required for a transition once a "proper" split happens...it just kinda sucks that we're essentially blocked here based on whether or not Jetbrains-infra's Vsphere plugin will get merged or not. Having to keep the namespace open for when (or if) is likely a pain too which leaves a dever having to hack functionality into the VMware plugin and Vsphere post-processor. Plus, "if" it really does end up getting merged, there will be a lot of issues (or complaints, heh) that will get reported based on the deprecation or renamed configuration options in the VMware builders..

So, in the future once a vsphere/esx plugin gets inherited, it's almost as if we'd need to develop some alias or warning/deprecation framework (like an extension of #6987) to handle the migration of people to the new builder.

I know you guys just had a release and so will be dealing with bug reports related to it for now, but is it worth splitting up the local/remote components of the VMware builder into its own unique builder? Or prior to that kind of "major" change, would it be better to develop some alias/deprecation component to help/guide users with the migration (like under `validate` so it's not mandatory perhaps)? Or maybe there's something else to do so that users can be ready for that sort of transition whenever it happens...

(sorry, probably should've created a new issue and crammed a bunch of issue #s into it)

**Comment by ajovedel** [comment]
@arizvisa I agree with you. Just going through the vSphere related code it looks like the local/remote building code is very much integrated and would require significant work to separate it. I think a fix like what @azr suggested in https://github.com/hashicorp/packer/pull/7076 works very well (at the very least it solves this issue)

**Comment by SwampDragons** [comment]
@azr's solution would have solved this, but caused issues for people who were using outpud_directory to set a local file.

So for this ticket my current thinking is: 

Short term solution to this problem is adding a separate option called remote_output_directory which is different than the local output_directory, which I'm happy to do for the next release.

But the long term solution is splitting the builders into `vmware-iso-esx`, `vmware-iso`, `vmware-vmx-esx`, and `vmware-vmx`. We have a half dozen Amazon builders; I think that model should work okay here too. It'll be a heavy refactor to do this, and because of the backwards incompatabilities in namespacing we'll need to save that for one of our minor releases (e.g. 1.4.0 or 1.5.0)

Neither solution really is relevant the "vsphere" namespace at all, though. We aren't holding that namespace open for jetbrains specifically so much as we're holding open for a builder that uses the vsphere api bindings instead of using esxcli.

**Comment by arizvisa** [comment]
@SwampDragons, ok cool. I'll work on splitting the VMware builder up into that separate ESX namespace as sort of an RFC/PR (as initially it should be mostly copy+paste work). But anyways, then the VMware Workstation drivers can slowly approach stability while we all work on proper support for ESX.

Then at some point we can try and figure out what's the proper way to deal with backwards compatibility when pulling the esx pieces out of the VMware builders.

**Comment by arizvisa** [comment]
Just a heads up,

I'm still working on this branch of splitting up of the builders up at [arizvisa/packer#vmware-esx.split](https://github.com/arizvisa/packer/tree/vmware-esx.split). I haven't created a PR yet because it's still kind of a "research" branch at the moment (imo).

But over the break I had to work on something unrelated in a project that I maintain and so I haven't had too much of a chance to finish this particular work on the VMware builders. I'll create a PR once I get a chance to revisit this so you guys can review+discuss, tho.

**Comment by azr** [comment]
Howdy @arizvisa, that's awesome. Any luck with your research ? 🙂Is it looking good ?

**Comment by arizvisa** [comment]
Development is slowly chugging along. I currently have the esx-conditionals from the vmware builders split up into a vmware-esx builder, and I think that the next thing I need to do is like remove steps from the vmware-esx builder that don't actually do anything (but since it was essentially a copy from the vmware builder). Eventually I'll need to figure out about re-using the common code between the vmware and vmware-esx builders too at some point. But yeah, slowly chugging along but I had to put it down for a bit due to my regular job, but I plan to pick it up again once I get that stage done.

I have a PR made so we can track/discuss it here: #7212. I also invited you as a collaborator so feel free to poke around at it if you want too.

Sorry, I realize now that this issue wasn't the best place to announce that I had started working on this.
